pkg/bundle/sync.go

```diff
@@ -509,18 +508,63 @@ func (b *bundle) patchResource(ctx context.Context, obj interface{}) error {
 // fields from the Update operation to the Apply operation. This is required
 // to ensure that the apply operations will also remove fields that were
 // created by the Update operation.
-func (b *bundle) migrateToApply(ctx context.Context, obj client.Object, subresource string) error {
-	managedFields := obj.GetManagedFields()
+func (b *bundle) migrateConfigMapToApply(ctx context.Context, obj client.Object, key string) (bool, error) {
+	// isOldConfigMapManagedFieldsEntry returns a function that checks if the given ManagedFieldsEntry is an old
+	// ConfigMap managed fields entry. We use this to check if we need to migrate the ConfigMap managed fields to
+	// the Apply field operation.
+	// Because oldFieldManager is not unique, we also check that the managed fields contains the data key we know was
+	// set by trust-manager (bundle.Status.Target.ConfigMap.Key).
+	isOldConfigMapManagedFieldsEntry := func(mf *metav1.ManagedFieldsEntry) bool {
+		if mf.Manager != oldFieldManager || mf.Operation != metav1.ManagedFieldsOperationUpdate || mf.Subresource != "" {
+			return false
+		}
+
+		if mf.FieldsV1 == nil || mf.FieldsV1.Raw == nil {
+			return false
+		}
 
+		var fieldset fieldpath.Set
+		if err := fieldset.FromJSON(bytes.NewReader(mf.FieldsV1.Raw)); err != nil {
+			return false // in case we cannot parse the fieldset, we assume it's not an old target
+		}
+
+		return fieldset.Has([]fieldpath.PathElement{
+			{
+				FieldName: ptr.To("data"),
+			},
+			{
+				FieldName: ptr.To(key),
+			},
+		})
+	}
+
+	needsUpdate := false
+	for _, mf := range obj.GetManagedFields() {
+		if !isOldConfigMapManagedFieldsEntry(&mf) {
+			continue
+		}
+		needsUpdate = true
+	}
+	if !needsUpdate {
+		return false, nil
+	}
+
+	var cm corev1.ConfigMap
+	if err := b.directClient.Get(ctx, client.ObjectKeyFromObject(obj), &cm); err != nil {
```

Comment:
Is this the reason for introducing an additional client (`directClient`)? My guess is that you are trying to avoid polluting the cache with all full configmaps in the cluster. Did you consider disabling cache for configmaps - to prevent the additional client, or will this not work?

Reply:
Maybe it can be done without caching all the full objects, but I just don't want to add the complexity of such an exception. Also, this migration can be a temporary solution that is removed after a few versions. I feel like this new uncached client is the most elegant solution (please let me know if you don't agree).

Reply:
Yeah, good point that this migration code is going to be deleted after a couple of releases. But with an additional client, the SSA code could be greatly simplified - if the additional client was a standard K8s client-go client - with support for apply confgurations. 😉